domains/pallets/messenger/src/migrations.rs

```diff
@@ -1,203 +1,213 @@
-    pub(super) type Outbox<T: Config> = CountedStorageMap<
-        Pallet<T>,
-        Identity,
-        (ChainId, ChannelId, Nonce),
-        Message<BalanceOf<T>>,
-        OptionQuery,
-    >;
-
-    pub(super) fn migrate_messenger_storages<T: Config>() -> Weight {
+    #[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+    pub(crate) struct WeightTags {
+        pub(crate) outbox: BTreeMap<(ChainId, MessageId), MessageWeightTag>,
+        pub(crate) inbox_responses: BTreeMap<(ChainId, MessageId), MessageWeightTag>,
+    }
+
+    pub(crate) fn migrate_message_weight_tags<T: Config>(count: u64) -> (u64, u64) {
         let mut reads = 0;
         let mut writes = 0;
-        let inbox_responses = InboxResponses::<T>::drain().collect::<Vec<_>>();
-        inbox_responses.into_iter().for_each(|(key, msg)| {
-            // we do one read from the old storage
-            reads += 1;
 
-            // we do one write to old storage and one write to new storage
-            writes += 2;
-
-            InboxResponsesNew::<T>::insert(key, msg);
-        });
+        reads += 1;
+        let Some(mut weight_tags) = MessageWeightTags::<T>::get() else {
+            // nothing to process, exit
+            return (reads, writes);
+        };
 
-        let outbox = Outbox::<T>::drain().collect::<Vec<_>>();
-        let mut outbox_count = BTreeMap::new();
-        outbox.into_iter().for_each(|(key, msg)| {
-            // we do one read from the old storage
-            reads += 1;
+        // migrate outbox weight tags
+        while writes < count {
+            match weight_tags.outbox.pop_first() {
+                // no more weight tags in outbox
+                None => break,
+                Some(((chain_id, message_id), weight_tag)) => {
+                    OutboxMessageWeightTags::<T>::insert((chain_id, message_id), weight_tag);
+                    writes += 1;
+                }
+            }
+        }
+
+        // migrate inbox responses weight tags
+        while writes < count {
+            match weight_tags.inbox_responses.pop_first() {
+                // no more weight tags in outbox
+                None => break,
+                Some(((chain_id, message_id), weight_tag)) => {
+                    InboxResponseMessageWeightTags::<T>::insert((chain_id, message_id), weight_tag);
+                    writes += 1;
+                }
+            }
+        }
+
+        writes += 1;
+        if weight_tags.outbox.is_empty() && weight_tags.inbox_responses.is_empty() {
+            MessageWeightTags::<T>::kill();
+        } else {
+            MessageWeightTags::<T>::set(Some(weight_tags));
+        }
+
+        (reads, writes)
+    }
 
-            // we do one write to old storage and one write to new storage
-            writes += 2;
+    pub(crate) fn remove_inbox_response_weight_tags<T: Config>(tags: Vec<(ChainId, MessageId)>) {
+        let mut maybe_weight_tags = MessageWeightTags::<T>::take();
```

Comment:
I think `MessageWeightTags::<T>::get()` should be enough, `take()` is `remove + get`